lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
+/*
+ * This sets or resets the resettable properties on the VmAgent instance.
+ */
+VmAgent.prototype.initializeProperties = function initializeProperties() {
     var self = this;
-    var log = this.log;
 
-    this.initializeEventWatcher();
+    if (self.watcher) {
+        self.watcher.stop();
+        self.watcher = null;
+    }
 
-    // Wrap our initial full sample in a retry-backoff
-    var opts = { uuid: this.uuid };
-    var fn = this.sendFullSample.bind(this, opts);
-    this.retryUpdate(fn, opts, onRetry);
+    // vasync doesn't allow us to clear a queue, so kill and create a new one.
+    if (self.queue) {
+        self.queue.kill();
+    }
 
-    function onRetry(err) {
-        if (err) {
-            log.error(err, 'Failed retry-backoff for initial sendFullSample');
+    self.queue = vasync.queue(function _updateVmapiVm(vm_uuid, callback) {
+        // closure so self is correct here
+        if (!self.ready) {
+            callback();
             return;
         }
+        self.updateVmapiVm(vm_uuid, callback);
+    }, 1);
 
-        log.info('Initial VMs state was successfully sent. Good to go');
-        self.eventWatcher.updateState(self.sample);
-        self.eventWatcher.lastCfgEvent = self.lastFullSample;
-        self.eventWatcher.start();
-    }
+    // set values to defaults
+    self.updateDelay = INITIAL_UPDATE_DELAY;
+    self.dirtyVms = [];
+    self.retryDelays = {};
+    self.ready = false;
+    self.lastSeenVms = {};
 };
 
-
-/*
- * Initializes the EventWatcher event listeners
- */
-VmAgent.prototype.initializeEventWatcher = function () {
+VmAgent.prototype.updateVmapiVm = function updateVmapiVm(vm_uuid, callback) {
     var self = this;
-    var log = this.log;
-    var eventWatcher = this.eventWatcher;
-
-    eventWatcher.on('state', function (uuid, state) {
-        log.debug('state event for %s state: %s', uuid, state);
-        self.pushSample({
-            uuid: uuid,
-            cachedVm: self.sample[uuid]
-        });
-    });
-
-    eventWatcher.on('zone_state', function (uuid, zone_state) {
-        log.debug('zone_state event for %s newstate: %s', uuid, zone_state);
-        self.pushSample({
-            uuid: uuid,
-            cachedVm: self.sample[uuid]
-        });
-    });
 
-    eventWatcher.on('zone_xml', function (uuid) {
-        log.debug('fs.watch event on /etc/zones for %s', uuid);
-        self.pushSample({
-            uuid: uuid,
-            cachedVm: self.sample[uuid]
-        });
-    });
-
-    eventWatcher.on('destroyed', function (uuid) {
-        var vm = self.sample[uuid];
-        if (!vm) {
-            log.warn('VM %s appears to have gone away but ' +
-                'self.sample doesn\'t have it', uuid);
-            return;
-        }
-
-        log.info('Found a destroyed VM %s', uuid);
-        vm.state = 'destroyed';
-        self.pushSample({ uuid: uuid, cachedVm: vm });
-    });
+    assert.ok(self.ready, 'no updates until init complete');
```

Comment:
Do we want to `assert` on this function's params (`vm_uuid` and `callback`)? That's usually what I do as it helps me through development more than in production. I would extend the question to any other function in the source.